### PR TITLE
Remove EnforceSingleRowNode support from PlanNodeDecorrelator

### PR DESCRIPTION
**Pull request by sopel39**
EnforceSingleRowNode guarantees that exactly single row is produced. Therefore, PlanNodeDecorrelator was incorrectly removing EnforceSingleRowNode even if decorrelated subplan could produce 0 rows.
Additionally, TransformCorrelatedJoinToJoin was completing with TransformCorrelatedScalarSubquery to decorrelate EnforceSingleRowNode.
This changes removes support for EnforceSingleRowNode from PlanNodeDecorrelator and makes TransformCorrelatedScalarSubquery the only rule that (correctly) decorrelates that node.

### CODE REVIEW

**Comment by sopel39** [comment]
This is mostly cleanup, I don't think it's possible to produce wrong plan with this issue.

**Comment by sopel39** [comment]
Fixes https://github.com/trinodb/trino/pull/19002

**Comment by findepi** [comment]
> Fixes #19002

This issue is closed.
Should it be reopened?

**Comment by findepi** [comment]
cc @kasiafi @kokosing for decorrelation

File: core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java [hunk 1]
```diff
@@ -268,7 +268,7 @@ private PlanBuilder planScalarSubquery(PlanBuilder subPlan, Cluster<SubqueryExpr
                 subPlan,
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java:268]: from commit message:

> Additionally, TransformCorrelatedJoinToJoin was completing
with TransformCorrelatedScalarSubquery to decorrelate
EnforceSingleRowNode.


did you mean  completing or competing ?

File: core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java [hunk 2]
```diff
@@ -268,7 +268,7 @@ private PlanBuilder planScalarSubquery(PlanBuilder subPlan, Cluster<SubqueryExpr
...
                 root,
                 scalarSubquery.getQuery(),
-                CorrelatedJoinNode.Type.LEFT,
+                CorrelatedJoinNode.Type.INNER,
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java:271]: perhaps this is obvious, but why this change?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java [hunk 3]
```diff
@@ -101,6 +102,7 @@ public Pattern<CorrelatedJoinNode> getPattern()
     @Override
     public Result apply(CorrelatedJoinNode correlatedJoinNode, Captures captures, Context context)
     {
+        checkArgument(correlatedJoinNode.getType() == INNER || correlatedJoinNode.getType() == LEFT, "unexpected correlated join type: " + correlatedJoinNode.getType());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java:105]: add a comment explaining why only INNER and LEFT are supposed to show up here and make it a `verify` (this is assertion-like check)

why isn't it a part of the PATTERN?

**Comment by sopel39** [comment]
> This issue is closed.
Should it be reopened?

@findepi this PR is the "proper" fix to the underlying issue (https://github.com/trinodb/trino/pull/19932 was not addressing the issue of bad planning covered by this PR)

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 4]
```diff
@@ -391,13 +390,6 @@ private Optional<OrderingScheme> decorrelateOrderingScheme(OrderingScheme orderi
             return Optional.of(new OrderingScheme(nonConstantOrderBy.build(), nonConstantOrderings.buildOrThrow()));
         }
 
-        @Override
-        public Optional<DecorrelationResult> visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
-        {
-            Optional<DecorrelationResult> childDecorrelationResultOptional = node.getSource().accept(this, null);
-            return childDecorrelationResultOptional.filter(result -> result.atMostSingleRow);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PlanNodeDecorrelator.java:398]: from cmt msg

> EnforceSingleRowNode guarantees that exactly single row
is produced. Therefore, PlanNodeDecorrelator was incorrectly
removing EnforceSingleRowNode even if decorrelated subplan
could produce 0 rows.

sounds like we could improve the logic (instead of removing it),
if we knew the plan produces exactly single row (eg global aggregation)

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestTransformCorrelatedScalarSubquery.java [hunk 5]
```diff
@@ -196,7 +201,8 @@ public void rewritesScalarSubquery()
                                 values("corr"),
                                 filter(
                                         "1 = a",
-                                        values("a"))));
+                                        values("a")))
+                                .with(CorrelatedJoinNode.class, join -> join.getType() == LEFT));
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestTransformCorrelatedScalarSubquery.java:205]: shouldn't `correlatedJoin` matcher support checking the join type?

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestTransformCorrelatedScalarSubquery.java [hunk 6]
```diff
@@ -186,6 +189,8 @@ public void rewritesScalarSubquery()
                 .on(p -> p.correlatedJoin(
                         ImmutableList.of(p.symbol("corr")),
                         p.values(p.symbol("corr")),
+                        INNER,
+                        TRUE_LITERAL,
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestTransformCorrelatedScalarSubquery.java:193]: Why this change?

if i revert it, the `correlatedJoin` method being called supplements INNER, TRUE_LITERAL by default
so it looks redundant to me

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java [hunk 7]
```diff
@@ -124,7 +126,10 @@ public Result apply(CorrelatedJoinNode correlatedJoinNode, Captures captures, Co
                     correlatedJoinNode.getInput(),
                     rewrittenSubquery,
                     correlatedJoinNode.getCorrelation(),
-                    producesSingleRow ? INNER : correlatedJoinNode.getType(),
+                    // EnforceSingleRowNode guarantees that exactly single matching row is produced
```
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java:129]: if there is no EnforceSingleRowNode, we exit earlier:

```
        if (!searchFrom(subquery, context.getLookup())
                .where(EnforceSingleRowNode.class::isInstance)
                .recurseOnlyWhen(ProjectNode.class::isInstance)
                .matches()) {
            return Result.empty();
```

[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java:129]: I missed that. It wasn't immediately visible in the diff. However, that should not be necessary for this rule to work in general, as far as I can tell. We omit that check and rely on the cardinality property of the subplan. 

I guess my comment is that in order to future-proof this comment from non-local changes, we could just generalize it to say that the subplan is guaranteed to produce one row (regardless of how that's achieved), so this can be an inner join instead of a left join.

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java:129]: I guess the rule insists on finding EnforceSingleRowNode because it also wants (needs?) to remove it.

**Review by martint — APPROVED** [review]
